### PR TITLE
Issue 6825: Update LTS Snapshot CLI command

### LINKED ISSUE #6825: Admin CLI for LTS Journal Snapshot updates/edits.
**Is your feature request related to a problem? Please describe.**
LTS journals today in Pravega hold metadata about the metadata segments that exist in Pravega. These metadata segments hold metadata about all non-metadata segments(both internal and user) in Pravega.
There can arise a need, where one might want to update these LTS journal snapshots. For e.g consider a situation where there has 
been some sort of data corruption that has happened on the metadata segments themselves, and to recover from this situation, lets say we use some recovery commands to help fix this corruption. One such command today is "table-segment recvoery" command in the Admin CLI.  On doing so, eventually one would need to update these new fixed metadata segment details. And we know that any sort of information about the metadata segment themselves is stored in the LTS journal files/snapshots. So one will need a way to update these journal shapshots saved in LTS. That would be the main motivation of this Admin CLI command.

**Describe the solution you'd like**
Provide a CLI command to help generate a LTS journal snapshot based on the segment chunks being pointed to in the command and also the latest journal snapshot consisting of the state of all the metadata segments.
- Input for the command be all the chunk files that make up one of the metadata segments whose state is stored in journals.
The command would sort these chunks based on their epochs and offsets and read them serially forming the chunk list and updating the segment properties. 
- Another input will be the journal file downloaded from tier-2 consisting of the latest snapshot state.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by abhinb**
Signed-off-by: Abhin Balur <Abhin_Balur@dell.com>

**Change log description**  
Provides a CLI command to update LTS snapshots. The CLI command generates a LTS snapshot file with details referenced by segment chunks given as command input. 
Full usage of the command is detailed below:-
Command: update-latest-journal-snapshot
Parameters:
1) segment-chunk-path : Directory where all segment chunks of the metadata segment to be updated are saved.
2) journal-path:  "Path where the latest journal file is stored.(same as latest-snapshot if the latest journal file happens to be a snapshot)
3) latest-snapshot: Path where the latest snapshot file is stored.
4) output-directory: Directory where the updated snapshot should be saved

**Purpose of the change**  
Fixes #6825 

**What the code does**  
Provides a CLI to generate a LTS snapshot file. Specifically update a segment within the snapshot file. 

**How to verify it**  
(Optional: steps to verify that the changes are effective)

### CODE REVIEW

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+
+/**
+ * This command in short helps us to update a storage journal snapshot file. Storage today stores metadata about the 4 special metadata segments
+ * in something called as journals. There are also snapshots of these journals created on reaching a fixed size and stored as a journal snapshot file. A snapshot is
+ * a self contained  state of all the 4 special metadata segments.
+ * There can be scnearios where one might want to change/update the state stored in these journal files. For e.g there could be situations where data stored in one of the
```
Comment (krishan-Rai) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java:46]: scnearios = scenarios

**Comment by sachin-j-joshi** [comment]
@abhinb This tool is too specific to FileSystem. What happens with ECS, S3 or other chunk storages?

**Comment by abhinb** [comment]
> @abhinb This implementation is too specific to FileSystem. What happens with ECS, S3 or other chunk storages? Also if you could move some of the heavy functionality to https://github.com/pravega/pravega/blob/master/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/UtilsWrapper.java That would be a better option.

@sachin-j-joshi  some of these recovery commands work offline i.e after we have downloaded the segment chunk files or journal snapshot files locally on any Pravega setup with Admin CLI up and running and CLI having access to these downloaded files. So in case of S3, i believe we would be downloading using the s3 browser the needed chunk data, which would be then available as files, to be fed to these commands. So from the command or PR perspective nothing changes, but yes the recovery procedure would follow different set of steps in case of S3 for example. The recovery procedure that talks about how to use this command to do a recovery is captured here https://github.com/pravega/pravega/pull/6831 and as of today it only covers FileSystem; but needs to cover other chunk storages as well. Please correct me if you see anything wrong here.

Sure. I will look at what can be moved to UtilsWrapper class.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+     * Creates a new instance of the StorageUpdateSnapshotCommand.
+     *
+     * @param args The arguments for the command.
+     */
+    public StorageUpdateSnapshotCommand(CommandArgs args) {
+        super(args);
+    }
+
+    @Override
+    public void execute() throws Exception {
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java:72]: sure.. will outline what execute does.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+            }
+        }
+    }
+
+    /**
+     * Helps determine the segment pointed to by the chunks refered in this command
+     * is the same as the one we have from journal file.
+     */
+    private boolean isSegmentBeingEdited(String segmentFromJournal, String segmentFromChunks) {
+        String[] segmentParts = segmentFromJournal.split("\\/");
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java:184]: Is this character a special one that needs to be defined as a constant too?

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        segmentFromJournal = segmentParts[segmentParts.length - 1];
+        if (segmentFromJournal.split(EPOCH_SPLITTER)[0].equalsIgnoreCase(segmentFromChunks.split(EPOCH_SPLITTER)[0])) {
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java:186]: You can just return `segmentFromJournal.split(EPOCH_SPLITTER)[0].equalsIgnoreCase(segmentFromChunks.split(EPOCH_SPLITTER)[0])`, as it is analogous as what you are doing here, right?

**Comment by RaulGracia** [comment]
@abhinb yes, makes sense. This PR takes a similar approach as #6753, in which we expect the data from the Pravega Cluster to be available for the tool locally (irrespective from what type of storage it is actually stored), and then we can perform and test the recovery procedure locally. Once we are ok with the "repaired data", is just a matter of overwriting the damaged data in the original cluster with the one generated by the CLI command.

**Review by RaulGracia — CHANGES_REQUESTED** [review]
Looks good, but I have some comments though. Question: has this command been verified on a real cluster in which we intentionally corrupt/delete an Attribute Index Segment chunk of Segment Metadata Table Segment?

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
+/**
+ * Copyright Pravega Authors.
+ * <p>
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java:3]: Please, copy/paste the exact header as in other files and above adding these `<p>` markers and other variations w.r.t the official header.

**Review by ShwethaSNayak — COMMENTED** [review]
LGTM, Is it possible to record **storage update-latest-journal-snapshot** command output and share?

File: cli/admin/src/test/java/io/pravega/cli/admin/segmentstore/StorageCommandsTest.java [hunk 6]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright Pravega Authors.
- *
+ * <p>
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/test/java/io/pravega/cli/admin/segmentstore/StorageCommandsTest.java]: updated with one from other files

File: cli/admin/src/test/java/io/pravega/cli/admin/segmentstore/StorageCommandsTest.java [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
+ * <p>
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/test/java/io/pravega/cli/admin/segmentstore/StorageCommandsTest.java]: changed.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+
+/**
+ * This command in short helps us to update a storage journal snapshot file. Storage today stores metadata about the 4 special metadata segments
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: done.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+ * in something called as journals. There are also snapshots of these journals created on reaching a fixed size and stored as a journal snapshot file. A snapshot is
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: added a link to SLTS PDP.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: changed to SLTS.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+ * a self contained  state of all the 4 special metadata segments.
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: done.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: done.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        final String latestSnapshot = getArg(2);
+        final String outputJournalPath = getArg(3);
+
+        File journalFile = new File(journalPath);
+        File latestSnapshotFile = new File(latestSnapshot);
+        Preconditions.checkState(journalFile.isFile(), "journal-path provided should point to a valid journal file");
+        Preconditions.checkState(latestSnapshotFile.isFile(), "snapshot file path provided should point to a valid file");
+        //check whether we are deserializing a snapshot file or a journal index file
+        JournalDeserializer deserializer = journalFile.getName().contains(SNAPSHOT) ? new JournalSnapshotDeserializer() : new JournalFileDeserializer();
+        //list all segment chunks sorted based on epoch and offset
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: done.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        final String segmentChunkPath = getArg(0);
+        final String journalPath = getArg(1);
+        final String latestSnapshot = getArg(2);
+        final String outputJournalPath = getArg(3);
+
+        File journalFile = new File(journalPath);
+        File latestSnapshotFile = new File(latestSnapshot);
+        Preconditions.checkState(journalFile.isFile(), "journal-path provided should point to a valid journal file");
+        Preconditions.checkState(latestSnapshotFile.isFile(), "snapshot file path provided should point to a valid file");
+        //check whether we are deserializing a snapshot file or a journal index file
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: ok. keep forgetting this. sure. done.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 3 above)
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: can be a constant. Changed to one.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 4 above)
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: right. changed.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 5 above)
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: used one from other files. Doesnt seem to have any <p> markers.

**Comment by abhinb** [comment]
> LGTM, Is it possible to record **storage update-latest-journal-snapshot** command output and share?

Sure. The output is just one snapshot file. So when the command executes successfully it just prints below which has the path where the snapshot file is generated.
```
Updating System Snapshot with relevant data.

SystemSnapshot Journal File has been created successfully at <output_path>
```

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: mentioning them now.

**Comment by abhinb** [comment]
> Looks good, but I have some comments though. Question: has this command been verified on a real cluster in which we intentionally corrupt/delete an Attribute Index Segment chunk of Segment Metadata Table Segment?
This is tested by intentionally corrupting Attribute Index Segment, making sure we see the 
"DataCorruptionException: BTreeIndex operation failed. Index corrupted. Wrong footer information"  and then trying the recovery procedure "https://github.com/pravega/pravega/pull/6831". 
But all this has been tried on locally stood up container based dev setup.

**Review by RaulGracia — COMMENTED** [review]
Minor comments.

**Comment by RaulGracia** [comment]
> But all this has been tried on locally stood up container based dev setup.

@abhinb and is this enough to validate the approach? Please, make the necessary tests as this PR seems close to be merged.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+
+/**
+ * This command helps us to update a storage journal snapshot file. SLTS today stores metadata about the 4 special metadata segments
+ * in something called as Journals.More about SLTS Journals here(https://github.com/pravega/pravega/wiki/PDP-34-(Simplified-Tier-2)#bootstrap). There are also snapshots of these journals created on reaching a fixed size and stored as a journal snapshot file. A snapshot is
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: done.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+ * a self contained  state of all the 4 special metadata segments ( storage_metadata, container_metadata and their corresponding attribute index segments)
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: corrected

[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: corrected

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+     *
+     * @param args The arguments for the command.
+     */
+    public StorageUpdateSnapshotCommand(CommandArgs args) {
+        super(args);
+    }
+
+    /**
+     * What this method does:-
+     * a) Reads chunks of the segment that needs to be updated
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: corrected

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+
+/**
+ * This command helps us to update a storage journal snapshot file. SLTS today stores metadata about the 4 special metadata segments
+ * in something called as Journals.More about SLTS Journals here(https://github.com/pravega/pravega/wiki/PDP-34-(Simplified-Tier-2)#bootstrap). There are also snapshots of these journals created on reaching a fixed size and stored as a journal snapshot file. A snapshot is
+ * a self contained  state of all the 4 special metadata segments ( storage_metadata, container_metadata and their corresponding attribute index segments)
+ * There can be scenarios where one might want to change/update the state stored in these journal files. For example, there could be situations where data stored in one of the
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: replaced with if.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: corrected

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+        final String segmentChunkPath = getArg(0);
+        final String journalPath = getArg(1);
+        final String latestSnapshot = getArg(2);
+        final String outputJournalPath = getArg(3);
+
+        File journalFile = new File(journalPath);
+        File latestSnapshotFile = new File(latestSnapshot);
+        Preconditions.checkState(journalFile.isFile(), "journal-path provided should point to a valid journal file");
+        Preconditions.checkState(latestSnapshotFile.isFile(), "snapshot file path provided should point to a valid file");
+        // Check whether we are deserializing a snapshot file or a journal index file
```
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: corrected

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/storage/StorageUpdateSnapshotCommand.java]: corrected

**Review by RaulGracia — APPROVED** [review]

**Comment by abhinb** [comment]
This PR has been tested now on Kubernetes cluster as well. As part of testing this a datacorruptionexception was introduced by manually editing a storage metadata segment. After which we should see a corruption in the logs of the form 

> Caused by: io.pravega.segmentstore.server.DataCorruptionException: BTreeIndex operation failed. Index corrupted.
	at io.pravega.segmentstore.server.attributes.SegmentAttributeBTreeIndex.handleIndexOperationException(SegmentAttributeBTreeIndex.java:808)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:986)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:970)
	... 9 common frames omitted
Caused by: io.pravega.common.util.IllegalDataFormatException: [AttributeIndex[3-11]] Wrong footer information. RootPage Offset (412160) + Length (396810) exceeds Footer Offset (3050).
	at io.pravega.common.util.btree.BTreeIndex.initialize(BTreeIndex.java:235)

Post which to recover the cluster, we use the `data-recovery tableSegment-recovery` command followed by the `update-latest-journal-snapshot` command following the steps outlined in https://github.com/pravega/pravega/pull/6831.

One observation while doing the tests was, we encountered a transient AttributeUpdate error briefly in the logs which looked like below.
`ERROR i.p.s.s.writer.AttributeAggregator - AttributeAggregator[5-23]: Unable to persist root pointer RootPointer=2754, LastSeqNo=7601.
io.pravega.segmentstore.contracts.BadAttributeUpdateException: [Segment '_system/containers/storage_metadata_5'] Bad Attribute Update (AttributeId = 80000000-0000-0000-0000-000000000007, UpdateType = ReplaceIfGreater, Value = 2754): Expected greater than '3140'`

As noted this was error was transient and did not affect the recovery as such and can be handled separately.

**Pull request merged**